batcher/sweep_batcher.go

```diff
@@ -0,0 +1,581 @@
+func (b *Batcher) Run(ctx context.Context) error {
+	runCtx, cancel := context.WithCancel(ctx)
+	defer cancel()
+
+	defer func() {
+		for _, batch := range b.batches {
+			batch.Wait()
+		}
+
+		b.wg.Wait()
+	}()
+
+	// First we fetch all the batches that are not in a confirmed state from
+	// the database. We will then resume the execution of these batches.
+	batches, err := b.FetchUnconfirmedBatches(runCtx)
+	if err != nil {
+		return err
+	}
+
+	for _, batch := range batches {
+		err := b.spinUpBatchFromDB(runCtx, batch)
+		if err != nil {
+			return err
+		}
+	}
+
+	for {
+		select {
+		case sweepReq := <-b.SweepReqs:
+			sweep, err := b.fetchSweep(runCtx, sweepReq)
+			if err != nil {
+				return err
+			}
+
+			err = b.handleSweep(runCtx, sweep, sweepReq.Notifier)
+			if err != nil {
+				return err
+			}
+
+		case id := <-b.completeChan:
+			delete(b.batches, id)
+
+		case err := <-b.errChan:
+			return err
+
+		case <-runCtx.Done():
+			return runCtx.Err()
+		}
+	}
+}
+
+// handleSweep handles a sweep request by either placing it in an existing
+// batch, or by spinning up a new batch for it.
+func (b *Batcher) handleSweep(ctx context.Context, sweep *Sweep,
+	notifier SpendNotifier) error {
+
+	// When batches purge sweeps they will be returned here in order to be
+	// handled again. In that case the batch will not define a notifier as
+	// it does not care about the future outcome of the sweep. The swap has
+	// already registered a notifier with the batcher.
+	if notifier != (SpendNotifier{}) {
+		b.wg.Add(1)
+		go b.monitorSpendAndNotify(ctx, sweep, notifier)
+	}
+
+	// Check if the sweep is already in a batch. If that is the case, we
+	// provide the sweep to that batch and return.
+	for _, batch := range b.batches {
+		if batch.sweepExists(sweep.SwapHash) {
+			return writeToBatchChan(ctx, batch, sweep)
+		}
+	}
+
+	// If one of the batches accepts the sweep, we provide it to that batch.
+	for _, batch := range b.batches {
+		if batch.AcceptsSweep(sweep) {
+			return writeToBatchChan(ctx, batch, sweep)
+		}
+	}
+
+	// If no batch is capable of accepting the sweep, we spin up a fresh
+	// batch and hand the sweep over to it.
+	batch, err := b.spinUpBatch(ctx)
+	if err != nil {
+		return err
+	}
+
+	return writeToBatchChan(ctx, batch, sweep)
+}
+
+// spinUpBatch spins up a new batch and returns it.
+func (b *Batcher) spinUpBatch(ctx context.Context) (*Batch, error) {
+	cfg := BatchConfig{
+		MaxTimeoutDistance: DefaultMaxTimeoutDistance,
+		BatchConfTarget:    defaultBatchConfTarget,
+	}
+
+	batchKit := BatchKit{
+		returnChan:       b.SweepReqs,
+		completeChan:     b.completeChan,
+		wallet:           b.wallet,
+		chainNotifier:    b.chainNotifier,
+		signerClient:     b.signerClient,
+		musig2SignSweep:  b.musig2ServerSign,
+		verifySchnorrSig: b.VerifySchnorrSig,
+		store:            b.store,
+	}
+
+	batch := NewBatch(cfg, batchKit)
+
+	id, err := batch.insertAndAcquireID(ctx)
```

Comment:
Why not do the insertion inside the `NewBatch` constructor?

Reply:
separation of concerns